h/views/api/helpers/links.py

```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+"""Helpers for generating and retrieving service link metadata"""
+
+from __future__ import unicode_literals
+
+
+def register_link(link, versions, registry):
+    """
+    Register an API service's metadata for its supported versions
+
+    Add an entry for the indicated ``link`` onto the ``registry`` for each
+    of the versions it claims to supported. These entries include basic
+    endpoint metadata like description and primary HTTP method.
+
+    These data structures are later referenced to deliver an index of our
+    API endpoints at `GET /api/`.
+
+    :arg link: Service metadata
+    :type link: dict
+    :arg versions: Supported versions (e.g. ``"v1"``)
+    :type versions: list(str)
+    :arg registry: Pyramid registry to put metadata in; mutated in place
+    :type registry: :class:`Pyramid.Registry`
+    """
+    # Get the HTTP method for use in the API links metadata
+    primary_method = link["method"]
+    if isinstance(primary_method, tuple):
+        # If the view matches multiple methods, assume the first one is
+        # preferred
+        link["method"] = primary_method[0]
+
+    if not hasattr(registry, "api_links"):
+        registry.api_links = {}
```

Comment:
We do this in a few places in the `h` app—adding attrs to `config.registry`. It has a bit of a smell to me and makes it hard to test...attributes showing up like magic is a little hard to keep under control...

Reply:
I can see that we do it in a handful of places but is adding arbitrary properties directly onto Pyramid's registry object something you're supposed to do?

Reply:
I followed the example of existing code in h when I added this originally. I'm not deeply familiar with idiomatic usage of the Pyramid registry so I don't know if there is a preferred alternative.

Reply:
> I can see that we do it in a handful of places but is adding arbitrary properties directly onto Pyramid's registry object something you're supposed to do?

I think it's not the greatest, which is why I pointed it out...but it's entrenched right now and I'm not going to choose this battle just yet...

I did look at Pyramid's `Registry` class docs yesterday and there isn't a great place to stick stuff. The only attribute it has that is that is "meant" for manipulation directly seems to be `settings`. Food for future consideration...

Reply:
My understanding is that `settings` is for configuration metadata pulled from env vars, config files and other sources, as opposed to data generated at runtime and exposed for use elsewhere in the app.